torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -577,6 +577,11 @@ void ProcessGroupNCCL::abortTimedOutCollectives(
         ncclComm->ncclCommAbort(exceptionMsg);
         abortedCommIds.emplace(buildNcclUniqueIdStr(ncclComm->getNcclId()));
       }
+      // Handle Exceptions on failed GPU operations by bringing the process
+      // down.
+      if (!terminateProcessGroup_.load()) {
+        work.handleNCCLGuard();
```

Comment:
This will not terminate the process since the exception will just be caught here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L585 and be logged.

Also, why do we need to check for `terminateProcessGroup_.load()`? One more thing, this can be done before all the abort logic right? Basically if the work is timed out we should just bring the process down.

Btw, don't we need to check for `asyncErrorHandling_` being true before we do this?

Reply:
> This will not terminate the process since the exception will just be caught here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L585 and be logged.

Good point, should we try to work around this by changing the handling there to let this exception through and crash? We can probably case it on some const string and check the error message for that string, and if it matches, crash the process instead of just logging it. 

> Btw, don't we need to check for asyncErrorHandling_ being true before we do this?

`abortTimedOutCollectives` is only called if `asyncErrorHandling_`, see https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L645

Reply:
> Also, why do we need to check for terminateProcessGroup_.load()? One more thing, this can be done before all the abort logic right? Basically if the work is timed out we should just bring the process down.

The current call to handleNCCLGuard in the workCleanupThread also checks for process group termination, I think the idea is that if concurrently the process group destructor has run and the PG is being destroyeed, there is no real need to bring down the process. cc @osalpekar for this understanding. 

Regarding still aborting NCCL communicators, shouldn't we still abort them as a "best practice" instead of leaving them unaborted? I guess it doesn't make too much of a difference though.